exasol_transformers_extension/udfs/question_answering_udf.py

```diff
@@ -0,0 +1,142 @@
+        """
+        result_df_list = []
+        for model_name in batch_df['model_name'].unique():
+            model_df = batch_df[batch_df['model_name'] == model_name]
+
+            if self.last_loaded_model_name != model_name:
+                self.set_cache_dir(model_df)
+                self.load_models(model_name)
+
+            model_pred_df = self.get_prediction(model_df)
+            result_df_list.append(model_pred_df)
+
+        result_df = pd.concat(result_df_list)
+        return result_df
+
+    def set_cache_dir(self, model_df: pd.DataFrame) -> None:
+        """
+        Set the cache directory in bucketfs of the specified model. Note that,
+        cache_dir class variable is used for testing purpose. This variable is
+        set to a local path only in unit tests.
+
+        :param model_df: The model dataframe to set the cache directory
+        """
+        model_name = model_df['model_name'].iloc[0]
+        bucketfs_conn_name = model_df['bucketfs_conn'].iloc[0]
+        sub_dir = model_df['sub_dir'].iloc[0]
+        bucketfs_location = bucketfs_operations.create_bucketfs_location(
+            self.exa.get_connection(bucketfs_conn_name))
+
+        model_path = bucketfs_operations.get_model_path(sub_dir, model_name)
+        self.cache_dir = bucketfs_operations.get_local_bucketfs_path(
+            bucketfs_location=bucketfs_location, model_path=str(model_path))
+
+    def load_models(self, model_name: str) -> None:
+        """
+        Load model and tokenizer model from the cached location in bucketfs
+
+        :param model_name: The model name to be loaded
+        """
+        self.last_loaded_model = self.base_model.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+        self.last_loaded_tokenizer = self.tokenizer.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+        self.last_loaded_model_name = model_name
+
+    def get_prediction(self, model_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform prediction of the given model and preparation of the prediction
+        results according to the format that the UDF can emit.
+
+        :param model_df: The dataframe to be predicted
+
+        :return: The dataframe where the model_df is formatted with the
+        prediction results
+        """
+        preds, labels = self._predict_model(model_df)
+        pred_df = self._prepare_prediction_dataframe(model_df, preds, labels)
+        return pred_df
+
+    def _predict_model(self, model_df: pd.DataFrame) -> \
+            Tuple[List[float], List[str]]:
+        """
+        Predict the given text list using recently loaded models, return
+        probability scores and labels
+
+        :param model_df: The dataframe to be predicted
+
+        :return: A tuple containing prediction score list and label list
+        """
+        questions = list(model_df['question'])
+        contexts = list(model_df['context_text'])
+        question_answerer = self.pipeline(
```

Comment:
I was able to extract the prediction score much more easily, that's why I prefer to use pipeline in here.
it would be better to use same api call for all predictions. Let's decide according to other tasks. If we can use the pipeline in others without any problems, we can use pipeline in them all.

Reply:
ok, good to know. Maybe mention this also in the refactoring ticket.